compiler/noirc_frontend/src/elaborator/mod.rs

```diff
@@ -1104,6 +1106,34 @@ impl<'context> Elaborator<'context> {
                 }
             } else {
                 for (_, func_id, _) in &overrides {
+                    let func_meta = self.interner.function_meta(func_id);
+                    for override_trait_constraint in func_meta.trait_constraints.clone() {
+                        // We allow where clauses on impls but during definition collection they are included as part of the where
+                        // clause of each function. This check is necessary so we do not unnecessarily error on trait impl where clauses.
+                        let override_constraint_is_from_impl =
+                            trait_impl_where_clause.iter().any(|impl_constraint| {
+                                impl_constraint.trait_id == override_trait_constraint.trait_id
+                            });
+                        if override_constraint_is_from_impl {
+                            continue;
+                        }
```

Comment:
We should still catch cases where the generics are different e.g.
```
trait Foo<T> {
    fn foo<U>() where T: Default;
}

impl<A> Foo<A> for () {
    fn foo<B>() where B: Default {}
}
```
I expect this'll be somewhat more code so we may want to break this `override_constraint_is_from_impl` check into another function as well.